### ISSUE TITLE
Extract libanki package into Android-free libanki jar

### ISSUE THREAD

**Issue by unoexperto** [ankidroid/Anki-Android#4517]
Hey guys,

I'd like to generate apkg files on the sever side. I was considering using libanki package but quick glance indicates that it heavily depends on android components. I can abstract those and extract entire package into separate project. Would you be interested in incorporating it later into master branch ? Or maybe I'm inventing a wheel and you already have solution for server-side ? My backend is in Scala.

Thanks!

**Comment by timrae** [comment]
I guess you want this:
https://github.com/dsnopek/anki-sync-server

**Comment by unoexperto** [comment]
Thanks! It's python though. I guess I'll take a look at anki-droid sources to estimate how long it would take to extract the package.

**Comment by timrae** [comment]
Ah right, sorry I didn't read your post properly. I'd need you to provide
detailed information on what you're trying to achieve and for what purpose
 in order to give any advice.

I don't think it'd be worth porting all of libanki to pure java, just the
parts you need. For example all the sync code would be a nightmare.

**Comment by mikehardy** [comment]
This is impractical for us - libanki is similar to the Python it came from but the amount of embedded special-case bugfixing is impressive.

**Comment by unoexperto** [comment]
@timrae What I'd love to have is relatively simple import/export format for decks. Something as simple as xml/json and media resourse files in zip file. Current format that uses SQLite is not very simple to produce. For example I'd like to create service that would use user provided list of words to produce media rich deck automatically (with mp3 and images).

**Comment by unoexperto** [comment]
@mikehardy What part exactly is impractical ?

**Comment by mikehardy** [comment]
Ripping out the android parts like Tim said from the sync code. But the importers and exporters... Should be possible? And would possibly improve the ecosystem by making them easier to develop so we could do a text one and a csv one (finally). I think you have a neat idea and this could work. I have maven repo namespace for libraries already even

**Comment by mikehardy** [comment]
I should mention that having a separate library of file/package <-> anki collection code might be a good idea even, as our most resource-intensive code (that is: uses most RAM, most disk space) is when we do collection import / export and could really use some performance-tuning. That sort of thing is usually easier in a nice tidy package

**Comment by unoexperto** [comment]
@mikehardy Brief code review indicates that despite having `com.ichi2.libanki.DB` class SQlite code spills over to other places such as `com.ichi2.libanki.Collection` and `com.ichi2.libanki.Models`. So I'd prefer to begin my work with incremental refactoring of existing code to separate model (in CS terms) classes and DB logic. And then abstract Android-only code such as `androidx.sqlite.*` or `android.text.TextUtils` into thin wrappers. Doing all thing in separate project and then integrating back to Anki-Android could be very error-prone. I'd rather refactor it within Anki-Android in small steps. Would you guys be comfortable accepting such pull-requests?

**Comment by mikehardy** [comment]
I am personally comfortable with that, but only if they are based on the tests added in my current database work from #5172 to make sure things are exercised, and with the idea that hopefully there are more tests to cover areas you change. Then on top of that, I would not do it without @timrae agreeing - he's got more AnkiDroid experience than I do by at least 10x and stops me from doing silly things before I do them (or at least before they are committed, usually)

**Comment by timrae** [comment]
It sounds ok in theory but tbh I don't see a big upside for us to accept these refactoring changes. It takes a lot of time to review PRs , and that's currently our development bottleneck. Also bugs will be inevitably introduced. So from my perspective, we spend a lot of time reviewing code and the best we hope to get from this is slightly less messy code with potentially more bugs. I'm not ruling it out in the future, but right now I think we have more important matters to focus on, sorry.

**Comment by unoexperto** [comment]
I spent this weekend looking at libanki package. I converted libanki package to non-Android state except DeckTask and "sync", "hooks" and "stats" packages. Here are my thoughts about it.

First of all, guys, I don't want to sound negative because I deeply appreciate what you do. Anki helped me and tons of my friends. But the way codebase is written is very very C-style. To name the few there are following issues (in my humble opinion):

- Manual parsing of JSON
- Too low-level DB operations
- Use of json objects as Java objects to store data in memory of the app
- Mixing business-logic and data operations
- Mixing UI calls into business/data layer
- Use of Java reflection
- Use of json array to store objects assuming relationship between "object" fields and array index
- Use of nested Maps instead of single Map with composite key
- Excessive use of flow-control continue-break

In long run it's non very maintainable code and it will (if not already) stall development of the product.

Here is initial result of stripping out Android-specific code: https://gitlab.com/unoexperto/libanki
It passes existing unit-tests for import/export functionality. I assumed your usernames are the same so I gave access to you on gitlab.

Now I'm thinking that it would be safer for me to continue refactoring within `Anki-Android` project than in separate project. I can do it in forked repo or I can do it in separate branch of your repo. Let me know what you prefer.

My further plan is this:

1) Replace Android SQLite with JDBC implementation. "fluentjdbc" wrapper + jdbc driver (xerial sqlite for JVM and SQLDroid for Android). Transactional integrity is temporary disabled;
2) Change visibility of methods of classes in libanki package to package visibility to limit exposure to UI code;
3) Replace use of JSONObject-s to store card collections (in Collection/Models/Card/Decks classes) with POJO classes;
4) Because of #3 move SQL-specific code to separate layer;
5) Define clear storage interface and hide SQLite implementation behind it;
6) Define interface for linanki classes to abstract logging and callbacks to UI.

Also after coding in Scala for last 6 years I feel very uneasy writing so much boiler-plate code in Java 8. So I considered option of using Lombok but after evaluating it I lean toward rewriting it in Kotlin. I intend to keep Java-style interfaces in the library for seamless integration into existing Anki-Droid. Anyway I won't code until step #6.

Let me know what you think.

**Comment by mikehardy** [comment]
@unoexperto - I think you found me? I'm @mikehardy.net on gitlab and I received an invite.

w.r.t. libanki and the way the code looks, that's because it is as direct a port from the Python as possible. As a Java port of a living Python project, fidelity with the original python (even if it is not a "clean" Java style or MVC or MVVC or whatever) is a feature.

Further, to get in front of Anki Desktop from the perspective of functionality is not necessarily useful when we are bound by those features for sync and interoperation etc.

So from that perspective, I'm not interested personally in libanki diverging much from the Python - us being "stalled" is just not having enough time to read the Python diffs and implement the #4905 v2 scheduler and #4915 MathJax for instance, it's not that we can't implement arbitrary new features. 

As of about 2 years, I think the project was "stalled" completely for lack of time, not code structure, and it took a few months simply to get project dependencies up to date for instance. Which is just to say I don't think the architecture (gnarly as it may be) is the limiter. In thinking of issues in the tracker I can't actually think of any where libanki's design are the problem? But I might be missing something.

If a refactor makes it easier to import/export decks or test things - that to me increases interoperability and I like the idea, that was the basis of my support earlier. But if it decreases the ability to ingest changes from the upstream Python by diverging a lot, that will likely decrease interoperability (as we diverge from Anki Desktop farther) and it makes me nervous.

With regard to Kotlin my thinking as a former team lead is "do what your team knows". I personally don't know what the contributors skills are inside AnkiDroid committers so I'm not sure if that works well or not. I don't know Kotlin myself so that'd be a barrier to me for contributions until I came up to speed.

If none of that works for you, maybe the best thing to do is to think of your project as a separate thing, consuming and producing artifacts in the Anki ecosystem, maybe using the AnkiDroid API to push changes into people's collections or maybe just making collections to download for import or similar?

**Comment by unoexperto** [comment]
> If a refactor makes it easier to import/export decks or test things - that to me increases interoperability and I like the idea, that was the basis of my support earlier. But if it decreases the ability to ingest changes from the upstream Python by diverging a lot, that will likely decrease interoperability (as we diverge from Anki Desktop farther) and it makes me nervous.

@mikehardy I see. Being as close as possible to Python makes sense to me. But it look like it means introducing new apkg format is not doable at all because logic in `Collection` and related classes are intertwined with SQLite calls. In other words unless Damien splits business logic from DB operations in his Python code you won't be able to port his commits to AnkiDroid. Am I correct?

I guess my only option then is to write apkg converter. Is there regularily updated documentation of format of apkg file (SQLite tables and format of serialzed json fields in it) ? Because column names like `ivl`, `crt` and etc. are not very self-explanatory.

**Comment by mikehardy** [comment]
> it look like it means introducing new apkg format is not doable at all because logic in `Collection` and related classes are intertwined with SQLite calls. In other words unless Damien splits business logic from DB operations in his Python code you won't be able to port his commits to AnkiDroid. Am I correct?

I believe you are correct. .apkg (or .colpkg) files are *the* interchange format amongst different parts of the ecosystem. They are sometimes downloaded directly (for full sync) inside the apps, and they are certainly passed around and archived etc by users to share decks. There hasn't been a need to change AFAICS because it appears to work very well over time, really low maintenance, 10 years of operation as a distributed system, and here we are with millions of users :-)

If you wanted to try to add a new format, I think you'd need to get it into the Desktop code first (it's on github at https://github.com/dae/anki/tree/master/anki, with propagation to the iOS and web and then AnkiDroid

> I guess my only option then is to write apkg converter. Is there regularily updated documentation of format of apkg file (SQLite tables and format of serialzed json fields in it) ? Because column names like `ivl`, `crt` and etc. are not very self-explanatory.

I think that is correct also - and at the same time, I think that could have a lot of value. As mentioned previously - for AnkiDroid even, it would be great to have the ability to ingest .csv files, or anything else - so for me if it was possible to do it within a library that could be used in Android that would be great, but even if it was just a separate project, it would help the ecosystem.

For structure, it's all in the wiki - https://github.com/ankidroid/Anki-Android/wiki/Database-Structure

**Comment by unoexperto** [comment]
@mikehardy Thank you for the wiki link! When I googled I found only [this link](http://decks.wikia.com/wiki/Anki_APKG_format_documentation) and it seemed dated. 

I'll stick to implementing format-to-format converter then. It looks less risky to me now. I assume convincing Damien to introduce new format would be even harder than convincing you :) I'll ping you guys when I'm done.

And if you ever need help with heavy refactoring let me know.
